### ISSUE TITLE
"useless-assign" rule reporting false positives

### ISSUE THREAD

**Issue by msorens** [returntocorp/sgrep-rules#1098]
**Describe the bug**
This rule: https://semgrep.dev/editor?registry=javascript.lang.correctness.useless-assign
reports findings on both of these and it should not:
```
// CASE A
// ok
q1 = result * 2;
q1 = q1 < 25 ? q1 : 0;

// CASE B
// ok
let progress = this.getProgress(file.name, percent);
progress = {...progress, loading: true};
```

**To Reproduce**
See above.

**Expected behavior**
Findings should not be reported for the two cases above.

**Additional Context**
Fixes:
(1) Unrelated, but there is a bug in the pre-existing tests in that rule.
The first few lines are not testing what was expected because of reusing a variable.
Should instead be
```
// ruleid:useless-assignment
var something_other_than_x = 1;

// ruleid:useless-assignment
x = 2;
x = 3;
console.log(x1);
```

(2) We can actually simplify the existing rule and still pass all the pre-existing tests with just this:
```
rules:
- id: useless-assignment
  patterns:

    - pattern: |
        $X = $Y;
        $X = $Z;

    - pattern-not: |
        $X = $Y;
        $X = <... $X ...>;       
```
This also covers my CASE A above but, alas, it does not cover CASE B, but I believe it should, so seems like a bug in the engine as well, perhaps.

Here it is in the playground: https://semgrep.dev/s/px90

**Comment by mschwager** [comment]
Hey @msorens,

I agree, this rule shouldn't be catching these false positives.

Regarding (1), can you elaborate here? I'm not sure I follow. In that case I wouldn't expect `something_other_than_x` to result in a finding.

Regarding (2), I like this simplification/improvement! I agree, I would expect `<... $X ...>` to include catching `{...progress, loading: true}`. I think this will likely have to be a change made in `semgrep-core`. cc @aryx does that seem reasonable?

**Comment by msorens** [comment]
For (1), the existing tests show this:
```
// ruleid:useless-assignment <<< CALL THIS TEST T1
var x = 1;

// ruleid:useless-assignment <<< CALL THIS TEST T2
x = 2;
x = 3;
console.log(x1);
```

T1 is affecting T2 and I do not think that was the intent.
T2 should be reporting a finding because of `x=3` but it is actually reporting a finding because of `x=2`. I think.

**Comment by mschwager** [comment]
Looks like T1 `ruleid` was explicitly added in https://github.com/returntocorp/semgrep-rules/pull/329. Looking at https://github.com/returntocorp/semgrep/pull/432, I think the `var x = 1;` was added to explicitly check for `var` definitions, not just bare assignments like `x = 2`. My interpretation would be that this was intentional, and that `x` was being uselessly assigned twice here. Thoughts?

EDIT: to be more clear here we could do something like:

```
// ruleid:useless-assignment
var x = 1;
x = 2;

// ruleid:useless-assignment
y = 1;
y = 2;
```

**Comment by msorens** [comment]
Ah, OK, that is fair, too.

**Comment by mschwager** [comment]
@msorens I created https://github.com/returntocorp/semgrep-rules/pull/1100 to fix this issue, and https://github.com/returntocorp/semgrep/issues/2552 to track improvements related to case B you noted above :+1:
